File: workbook/ch08/data/btree.c
#include <stdio.h>
#include <stdlib.h>

#define T 2  // minimum degree (order of tree)

typedef struct BTreeNode {
    int keys[2 * T - 1]; 
    struct BTreeNode *children[2 * T];
    int num_keys;  
    int leaf;      
} BTreeNode;

typedef struct {
    BTreeNode *root;
} BTree;

BTreeNode *createNode(int leaf) {
    BTreeNode *node = (BTreeNode *)malloc(sizeof(BTreeNode));
    node->num_keys = 0;
    node->leaf = leaf;
    for (int i = 0; i < 2 * T; i++)
        node->children[i] = NULL;
    return node;
}

BTreeNode *search(BTreeNode *node, int key) {
    if (node == NULL) return NULL;
    int i = 0;
    while (i < node->num_keys && key > node->keys[i])
        i++;

    if (i < node->num_keys && key == node->keys[i])
        return node;

    if (node->leaf)
        return NULL;

    return search(node->children[i], key);
}

void splitChild(BTreeNode *parent, int i) {
    BTreeNode *node = parent->children[i];
    BTreeNode *new_node = createNode(node->leaf);
    new_node->num_keys = T - 1;

    for (int j = 0; j < T - 1; j++)
        new_node->keys[j] = node->keys[j + T];

    if (!node->leaf) {
        for (int j = 0; j < T; j++)
            new_node->children[j] = node->children[j + T];
    }

    node->num_keys = T - 1;

    for (int j = parent->num_keys; j >= i + 1; j--)
        parent->children[j + 1] = parent->children[j];

    parent->children[i + 1] = new_node;

    for (int j = parent->num_keys - 1; j >= i; j--)
        parent->keys[j + 1] = parent->keys[j];

    parent->keys[i] = node->keys[T - 1];
    parent->num_keys++;
}

void insertNonFull(BTreeNode *node, int key) {
    int i = node->num_keys - 1;

    if (node->leaf) {
        while (i >= 0 && key < node->keys[i]) {
            node->keys[i + 1] = node->keys[i];
            i--;
        }
        node->keys[i + 1] = key;
        node->num_keys++;
    } else {
        while (i >= 0 && key < node->keys[i])
            i--;
        i++;

        if (node->children[i]->num_keys == (2 * T - 1)) {
            splitChild(node, i);
            if (key > node->keys[i])
                i++;
        }
        insertNonFull(node->children[i], key);
    }
}

void insert(BTree *tree, int key) {
    BTreeNode *root = tree->root;
    if (root->num_keys == (2 * T - 1)) {
        BTreeNode *new_root = createNode(0);
        new_root->children[0] = root;
        tree->root = new_root;
        splitChild(new_root, 0);
    }
    insertNonFull(tree->root, key);
}

void traverse(BTreeNode *node) {
    if (node == NULL) return;
    for (int i = 0; i < node->num_keys; i++) {
        if (!node->leaf)
            traverse(node->children[i]);
        printf("%d ", node->keys[i]);
    }
    if (!node->leaf)
        traverse(node->children[node->num_keys]);
}

int main() {
    BTree tree;
    tree.root = createNode(1);

    int keys[] = {10, 20, 5, 6, 12, 30, 7, 17};
    for (int i = 0; i < 8; i++)
        insert(&tree, keys[i]);

    printf("B-Tree traversal:\n");
    traverse(tree.root);
    printf("\n");

    return 0;
}

File: workbook/ch07/data/btree.py
class BTreeNode:
    def __init__(self, t, leaf=False):
        self.t = t  # Minimum degree (defines range for number of keys)
        self.leaf = leaf  # True if leaf node
        self.keys = []  # List of keys
        self.children = []  # List of children

    def __str__(self):
        return str(self.keys)

class BTree:
    def __init__(self, t):
        self.root = BTreeNode(t, True)
        self.t = t

    def search(self, key, node=None):
        if node is None:
            node = self.root

        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1

        if i < len(node.keys) and node.keys[i] == key:
            return node

        if node.leaf:
            return None

        return self.search(key, node.children[i])

    def insert(self, key):
        root = self.root

        if len(root.keys) == (2 * self.t) - 1:
            new_root = BTreeNode(self.t, False)
            new_root.children.append(root)
            self.split_child(new_root, 0)
            self.root = new_root

        self.insert_non_full(self.root, key)

    def insert_non_full(self, node, key):
        i = len(node.keys) - 1

        if node.leaf:
            node.keys.append(0)
            while i >= 0 and key < node.keys[i]:
                node.keys[i + 1] = node.keys[i]
                i -= 1
            node.keys[i + 1] = key
        else:
            while i >= 0 and key < node.keys[i]:
                i -= 1
            i += 1

            if len(node.children[i].keys) == (2 * self.t) - 1:
                self.split_child(node, i)
                if key > node.keys[i]:
                    i += 1

            self.insert_non_full(node.children[i], key)

    def split_child(self, parent, i):
        t = self.t
        node = parent.children[i]
        new_node = BTreeNode(t, node.leaf)
        
        parent.keys.insert(i, node.keys[t - 1])
        parent.children.insert(i + 1, new_node)

        new_node.keys = node.keys[t:(2 * t) - 1]
        node.keys = node.keys[0:t - 1]

        if not node.leaf:
            new_node.children = node.children[t:(2 * t)]
            node.children = node.children[0:t]

    def traverse(self, node=None):
        if node is None:
            node = self.root
        for i in range(len(node.keys)):
            if not node.leaf:
                self.traverse(node.children[i])
            print(node.keys[i], end=" ")
        if not node.leaf:
            self.traverse(node.children[len(node.keys)])

# example
btree = BTree(2)
for key in [10, 20, 5, 6, 12, 30, 7, 17]:
    btree.insert(key)

print("B-Tree traversal:")
btree.traverse()
print()

print("Search for 12:", "Found" if btree.search(12) else "Not Found")
print("Search for 99:", "Found" if btree.search(99) else "Not Found")
